pkg/agent/core/metrics.go

```diff
@@ -0,0 +1,182 @@
+//
+// In general, there's only *really* one ScalingAlgorithm implementation, but we keep it generic so
+// that tests can dependency-inject an alternate implementation that's more
+
+import (
+	"fmt"
+	"math"
+	"strconv"
+	"strings"
+
+	"github.com/neondatabase/autoscaling/pkg/api"
+	"github.com/neondatabase/autoscaling/pkg/util"
+)
+
+type Metrics struct {
+	LoadAverage1Min  float32
+	LoadAverage5Min  float32 // unused. To be removed when api.Metrics.LoadAverage5Min is removed.
+	MemoryUsageBytes float32
+}
+
+// ToAPI converts the Metrics to the subset used for communications with the scheduler plugin and
+// exposed in pkg/api
+//
+// ToAPI implements ToAPIMetrics.
+func (m Metrics) ToAPI() api.Metrics {
+	return api.Metrics{
+		LoadAverage1Min:  m.LoadAverage1Min,
+		LoadAverage5Min:  m.LoadAverage5Min,
+		MemoryUsageBytes: m.MemoryUsageBytes,
+	}
+}
+
+// ScalingAlgorithm is the abstract interface provided by the default scaling algorithm, so that a
+// simpler version can be dependency-injected for easier testing.
+//
+// And also, separation of concerns is nice. Where state.go (and State within it) are more concerned
+// with the overall process of scaling approval and notification by communicating with other
+// components, the implementation(s) of ScalingAlgorithm can be more focused, making the calculation
+// of desired state abundantly clear.
+type ScalingAlgorithm[M ToAPIMetrics] interface {
+	// Update adds recent metrics
+	Update(metrics M)
+
+	// GoalCU returns the desired number of compute units that the VM should be using, given data
+	// from prior calls to Update and the configured resources equal to one compute unit.
+	GoalCU(scalingConfig api.ScalingConfig, computeUnit api.Resources) (_ uint32, ok bool)
+
+	// DeepCopy produces a copy of the value, intended for use when serializing State
+	DeepCopy() ScalingAlgorithm[M]
+}
+
+// ToApiMetrics is the interface that ScalingAlgorithm's metrics must satisfy - namely, that they
+// can be converted to an api.Metrics
+type ToAPIMetrics interface {
+	ToAPI() api.Metrics
+}
+
+// SimpleFactorScaling is a ScalingAlgorithm implementation that calculates the desired resource
+// usage by a simple multiple of the most recently collected metrics
+type SimpleFactorScaling struct {
+	// nb: fields only public to allow JSON serialization
+	Metrics *Metrics
+}
+
+func NewSimpleFactorScaling() *SimpleFactorScaling {
+	return &SimpleFactorScaling{
+		Metrics: nil,
+	}
+}
+
+// DeepCopy implements ScalingAlgorithm[Metrics]
+func (s *SimpleFactorScaling) DeepCopy() ScalingAlgorithm[Metrics] {
+	return &SimpleFactorScaling{
+		Metrics: shallowCopy[Metrics](s.Metrics),
+	}
+}
+
+// Update implements ScalingAlgorithm[Metrics]
+func (s *SimpleFactorScaling) Update(metrics Metrics) {
+	s.Metrics = &metrics
+}
+
+// GoalCU implements ScalingAlgorithm[Metrics], refer there for more.
+func (s *SimpleFactorScaling) GoalCU(scalingConfig api.ScalingConfig, computeUnit api.Resources) (_ uint32, ok bool) {
```

Comment:
Happy to explain — the idea behind these return values is that naming one of them `_` prevents naked returns (which, yeah are generally not great), while allowing us to provide names associated with the return values.

Otherwise, IMO it's hard to tell what the return values mean from `func GoalCU(...) (uint32, bool)`.

Reply:
Having named return values without using naked return is not uncommon (source: [Go stdlib](https://cs.opensource.google/go/go/+/refs/tags/go1.21.6:src/sync/map.go;l=330)).

In this particular case, I think having `(_ uint32, ok bool)` doesn't give us any more info. I think a comment for clarification would be good.